synapse/axon.py

```diff
@@ -1,801 +1,1153 @@
 import synapse.common as s_common
+import synapse.daemon as s_daemon
 import synapse.eventbus as s_eventbus
+import synapse.telepath as s_telepath
 
 import synapse.lib.cell as s_cell
+import synapse.lib.coro as s_coro
 import synapse.lib.lmdb as s_lmdb
 import synapse.lib.const as s_const
 
 logger = logging.getLogger(__name__)
 
-zero64 = b'\x00\x00\x00\x00\x00\x00\x00\x00'
-blocksize = s_const.mebibyte * 64
+# File convention: bsid -> unique id for a blobstor (actually the cell iden)
 
-class BlobStor(s_eventbus.EventBus):
+
+CHUNK_SIZE = 16 * s_const.mebibyte
+
+async def to_aiter(it):
+    '''
+    Take either a sync or async iteratable and yields as an async generator
     '''
-    The blob store maps buid,indx values to sequences of bytes stored in a LMDB database.
+    if hasattr(it, '__aiter__'):
+        async for i in it:
+            yield i
+    else:
+        for i in it:
+            yield i
+
+def _find_hash(curs, key):
     '''
+    Set a LMDB cursor to the first key that starts with \a key
 
-    def __init__(self, dirn, mapsize=s_const.tebibyte):
+    Returns:
+        False if key not present, else True and positions cursor at first chunk of value
+    '''
+    if not curs.set_range(key):
+        return False
+    return curs.key()[:len(key)] == key
 
-        s_eventbus.EventBus.__init__(self)
+class PassThroughApi(s_cell.CellApi):
+    '''
+    Class that passes through methods made on it to its cell.
+    '''
+    allowed_methods = []
 
-        path = s_common.gendir(dirn, 'blobs.lmdb')
-        self.lenv = lmdb.open(path, writemap=True, max_dbs=128)
-        self.lenv.set_mapsize(mapsize)
+    def __init__(self, cell, link):
+        s_cell.CellApi.__init__(self, cell, link)
 
-        self._blob_info = self.lenv.open_db(b'info')
-        self._blob_bytes = self.lenv.open_db(b'bytes') # <fidn><foff>=<byts>
+        for f in self.allowed_methods:
+            # N.B. this curious double nesting is due to Python's closure mechanism (f is essentially captured by name)
+            def funcapply(f):
+                def func(*args, **kwargs):
+                    return getattr(cell, f)(*args, **kwargs)
+                return func
+            setattr(self, f, funcapply(f))
+
+class IncrementalTransaction(s_eventbus.EventBus):
+    '''
+    An lmdb write transaction that commits if the number of outstanding bytes to be commits grows too large.
+
+    Naturally, this breaks transaction atomicity.
+    '''
+    MAX_OUTSTANDING = s_const.gibibyte
 
-        self._blob_clone = s_lmdb.Seqn(self.lenv, b'clone')
-        self._blob_metrics = s_lmdb.Metrics(self.lenv)
+    def __init__(self, lenv):
+        s_eventbus.EventBus.__init__(self)
+        self.lenv = lenv
+        self.txn = None
+        self._bytecount = 0
 
         def fini():
-            self.lenv.sync()
+            self.commit()
+            # lmdb workaround:  close environment from here to avoid lmdb crash
             self.lenv.close()
```

Comment:
I was going to ask - given multiple things could have access to the lenv, who else may be responsible for closing the lenv?

Reply:
there shouldn't be anybody else.